Forward AD formulas for activation backwards

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70468
* #70355
* __->__ #70460
* #69956

Differential Revision: [D33405363](https://our.internmc.facebook.com/intern/diff/D33405363)

soulitzer committed:
Forward AD formulas for activation backwards

[ghstack-poisoned]

soulitzer committed:
Update on "Forward AD formulas for activation backwards"




[ghstack-poisoned]

soulitzer committed:
Update on "Forward AD formulas for activation backwards"




[ghstack-poisoned]

soulitzer committed:
Update on "Forward AD formulas for activation backwards"




[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1791,6 +1792,7 @@
 - name: hardshrink_backward(Tensor grad_out, Tensor self, Scalar lambd) -> Tensor
   grad_out: hardshrink_backward(grad, self, lambd)
   self: zeros_like(grad)
+  result: at::where((self_p > lambd).logical_or(self_p < -lambd), grad_out_t, zeros_like(result))
```
No need to generate a full Tensor of 0s here?
Also `self_t` is not involved here?

soulitzer replied:
> No need to generate a full Tensor of 0s here?

Fixed

> Also self_t is not involved here?

Yeah the values of self are only used to compute the condition for where. Otherwise hardshrink backward is constant wrt self.

soulitzer committed:
Update on "Forward AD formulas for activation backwards"




[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405363).

albanD reviewed (approved):
SGTM

soulitzer committed:
Update on "Forward AD formulas for activation backwards"


Differential Revision: [D33405363](https://our.internmc.facebook.com/intern/diff/D33405363)

[ghstack-poisoned]

soulitzer committed:
Update on "Forward AD formulas for activation backwards"


Differential Revision: [D33405363](https://our.internmc.facebook.com/intern/diff/D33405363)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405363).

soulitzer committed:
Update on "Forward AD formulas for activation backwards"


Differential Revision: [D33405363](https://our.internmc.facebook.com/intern/diff/D33405363)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405363).

soulitzer committed:
Update on "Forward AD formulas for activation backwards"


Differential Revision: [D33405363](https://our.internmc.facebook.com/intern/diff/D33405363)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405363).
